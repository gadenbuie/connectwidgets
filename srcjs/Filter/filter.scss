.rscfilter {
  display: flex;
  position: relative;
  line-height: 3em;

  & button:focus {
    outline: none;
  }

  &__toggle {
    background: transparent;
    border: none;
    cursor: pointer;
    display: inline-block;
    font-size: 1.2em;
    line-height: 3em;
    padding: 0;
    padding-right: 2em;
    position: relative;
    text-align: left;
  }
  
  &__toggle::after {
    background-image: url("data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTAiIGhlaWdodD0iNSIgdmlld0JveD0iMCAwIDEwIDUiIGZpbGw9Im5vbmUiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+CjxwYXRoIGQ9Ik0wIC00LjM3MTE0ZS0wN0w1IDVMMTAgMEwwIC00LjM3MTE0ZS0wN1oiIGZpbGw9IiM0NDQ0NDQiLz4KPC9zdmc+Cgo=");
    background-position: center center;
    background-repeat: no-repeat;
    content: " ";
    height: 3em;
    position: absolute;
    right: 0;
    top: 0;
    width: 2em;
  }

  &__count {
    margin-left: 5px;
    /* more default styles on theming files */

    &:empty {
      display: none;
    }
  }
}
