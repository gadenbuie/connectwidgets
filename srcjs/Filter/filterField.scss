.rscfilter-field {
  position: relative;

  &__input-wrap{
    align-items: center;
    display: flex;
    position: relative;
  }

  &--dropdown::after {
    background-image: url("data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTAiIGhlaWdodD0iNSIgdmlld0JveD0iMCAwIDEwIDUiIGZpbGw9Im5vbmUiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+CjxwYXRoIGQ9Ik0wIC00LjM3MTE0ZS0wN0w1IDVMMTAgMEwwIC00LjM3MTE0ZS0wN1oiIGZpbGw9IiM0NDQ0NDQiLz4KPC9zdmc+Cgo=");
    background-position: center center;
    background-repeat: no-repeat;
    content: " ";
    height: 2.4em;
    position: absolute;
    right: 0;
    top: 0;
    width: 2.4em;
  }

  &__name {
    margin: 0;
    width: 30%;
  }

  &__input {
    box-sizing: border-box;
    height: 2.4em;
    padding: 0 0.7em;
    width: 70%;

    &:focus {
      outline: none;
    }
  }

  &__options {
    box-sizing: border-box;
    opacity: 0;
    pointer-events: none;
    max-height: 12em;
    left: 30%;
    overflow-y: auto;
    padding: 0.7em;
    position: absolute;
    top: 2.4em;
    width: 70%;
    z-index: 1;
  }

  &:focus-within &__options {
    opacity: 1;
    pointer-events: auto;
  }
}


