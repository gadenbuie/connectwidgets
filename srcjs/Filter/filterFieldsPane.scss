.rscfilter-pane {
  border-radius: 3px;
  box-shadow: 0 0 3px 1px rgba(0,0,0, 0.2);
  box-sizing: border-box;
  left: 0;
  line-height: 2.4em;
  opacity: 0;
  pointer-events: none;
  padding: 1em;
  position: absolute;
  top: 3em;
  transform: translateY(-1em);
  transition: all 0.3s ease;
  width: 360px;
  z-index: 100;

  /* Overlay-box's arrow */
  &::before {
    box-shadow: -1px -1px 1px 0px rgba(0, 0, 0, 0.06);
    content: " ";
    height: 1em;
    left: 1em;
    position: absolute;
    top: -0.4em;
    transform: rotate(45deg);
    width: 1em;
  }

  &--active {
    opacity: 1;
    pointer-events: auto;
    transform: translateY(0.5em);
  }
  
  &__actions {
    padding-top: 15px;
    text-align: right;

    /*
    &-reset {
      default styles on theming files
    }
    */
  }  
}
