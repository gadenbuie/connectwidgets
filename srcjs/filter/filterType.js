/**
 * Filter types
 */
 export const FilterType = {
  owner: 'owner',
  type: 'type',
  tag: 'tag',
};
