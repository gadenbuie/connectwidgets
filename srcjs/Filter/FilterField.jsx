import PropTypes from 'prop-types';
import { debounce } from '@/utils';
import FilterOptions from './FilterOptions';
import FiltersSelected from './FiltersSelected';
import { FilterType } from './filterType';
import { FilterContext } from './stateUtils';
import './filterField.scss';

const fieldClass = {
  base: 'rscfilter-field__input-wrap',
  dropdown: 'rscfilter-field__input-wrap rscfilter-field__input-wrap--dropdown',
};

function FilterField({ label, placeholder, type }) {
  const inputRef = React.createRef();

  const context = React.useContext(FilterContext);
  const isDropdown = type === FilterType.type;

  const prefilteringHandler = ev => {
    const prefix = ev.target.value.trim();
    context.searchFieldOptions(type, prefix);
  }

  const handleEsc = ev => {
    if (ev.key === "Escape") {
      ev.stopPropagation();
      inputRef.current.blur();
    }
  }

  return (
    <div className="rscfilter-field">
      <div className={isDropdown ? fieldClass.dropdown : fieldClass.base}>
        <label className="rscfilter-field__name">{label}</label>
        <input
          className="rscfilter-field__input"
          ref={inputRef}
          name={type}
          placeholder={placeholder}
          onChange={debounce(300, prefilteringHandler)}
          onKeyUp={handleEsc}
        />
        <FilterOptions type={type} />
      </div>
      <FiltersSelected type={type} />
    </div>
  );
}

FilterField.propTypes = {
  type: PropTypes.string,
  label: PropTypes.string,
  placeholder: PropTypes.string,
}

export default FilterField;
