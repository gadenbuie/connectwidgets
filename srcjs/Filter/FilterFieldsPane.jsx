import PropTypes from 'prop-types';
import FilterField from './FilterField';
import { FilterType } from './filterType';
import { FilterContext } from './stateUtils';
import './filterFieldsPane.scss';

function FilterFieldsPane({ isActive }) {
  const context = React.useContext(FilterContext);

  let paneClass = 'rscfilter-pane bg-white';

  if (isActive) {
    paneClass += ' rscfilter-pane--active';
  }

  return (
    <div className={paneClass}>
      <FilterField
        label="Owner"
        placeholder="Filter people"
        type={FilterType.owner}
      />
      <FilterField
        label="Type"
        placeholder="Select content types"
        type={FilterType.type}
      />
      <FilterField
        label="Tags"
        placeholder="Filter tags"
        type={FilterType.tag}
      />
      <div className="rscfilter-pane__actions border-top">
        <button
          className="rscfilter-pane__actions-reset btn btn-secondary"
          onClick={() => context.resetFilters()}
        >
          Reset
        </button>
      </div>
    </div>
  );
}

FilterFieldsPane.propTypes = {
  isActive: PropTypes.bool,
}

export default FilterFieldsPane;
