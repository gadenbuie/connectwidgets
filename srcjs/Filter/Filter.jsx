/**
 * rscfilter main component
 * React provided at window via reactR
 */

import PropTypes from 'prop-types';
import { cloneDeep } from 'lodash';
import { FilterType } from './filterType';
import FilterFieldsPane from './FilterFieldsPane';
import { 
  FilterContext,
  parseDataToFilters,
  shallowFiltersState,
} from './stateUtils';
import './filter.scss';

class Filter extends React.Component {
  constructor(props) {
    super(props);

    const crosstalkHandle = new crosstalk.FilterHandle();
    crosstalkHandle.setGroup(props.crosstalkGroup);

    const dataKeys = Array.isArray(props.crosstalkKey)
                      ? props.crosstalkKey
                      : props.data[props.crosstalkKey];

    this.parsedData = parseDataToFilters(props.data);
    this.state = {
      showPane: false,
      crosstalkHandle,
      dataKeys,
      options: cloneDeep(this.parsedData),
      selections: shallowFiltersState(),
      stateOptionsFor: this.stateOptionsFor.bind(this),
      stateSelectionsFor: this.stateSelectionsFor.bind(this),
      searchFieldOptions: this.searchFieldOptions.bind(this),
      updateSelection: this.updateSelection.bind(this),
      resetFilters: this.resetFilters.bind(this),
    };

    this.togglePane = this.togglePane.bind(this);
    this.setupCloseListeners();
  }

  setupCloseListeners() {
    const toggleIfNeeded = ev => {
      const withEscKey = ev.key && ev.key === 'Escape';
      const outOfPane = !ev.target.className.includes('rscfilter');
      if (
        this.state.showPane &&
        (withEscKey || outOfPane)
      ) {
        this.togglePane();
      }
    };

    document.addEventListener('keyup', toggleIfNeeded);
    document.addEventListener('click', toggleIfNeeded);
  }

  stateOptionsFor(type) {
    return this.state.options[type];
  }

  stateSelectionsFor(type) {
    return this.state.selections[type];
  }

  hasFiltersSelected() {
    const anyActiveSelection = Object.values(FilterType)
                                     .find(type => this.state.selections[type].length);
    return Boolean(anyActiveSelection);
  }

  getFiltersMatchingRows() {
    let matchingRows = [];

    Object.values(FilterType).forEach(filterType => {
      const selections = this.state.selections[filterType];
      const applicableRows = [
        ...new Set(selections.reduce((rows, sel) => [...rows, ...sel.applicableRows], []))
      ];
      if (selections.length) {
        if (!matchingRows.length) {
          matchingRows.push(...applicableRows)
        } else {
          matchingRows = matchingRows.filter(row => applicableRows.includes(row))
        }
      }
    });

    return matchingRows.map(rowIndex => this.state.dataKeys[rowIndex]);
  }

  searchFieldOptions(type, prefix) {
    let newOptionsState = cloneDeep(this.parsedData[type]);
    const currentSelections = this.state.selections[type];

    // If search is empty and there are pre-filtered options, reset and mark those selected
    if (!prefix && currentSelections.length) {
      currentSelections.forEach(opt => newOptionsState[opt.index].isSelected = true);
    } else {
    // Else, filter options with prefix and mark those already selected
      const reg = new RegExp(prefix, "i");
      newOptionsState = newOptionsState.filter(opt => reg.test(opt.label)).map(opt => {
        opt.isSelected = Boolean(currentSelections.find(s => s.name === opt.name));
        return opt;
      });
    }
    
    this.setState({
      options: { ...this.state.options, [type]: newOptionsState }
    });
  }

  updateSelection(target) {
    const { name, isSelected, filterType } = target;
    const newOptionsState = cloneDeep(this.state.options[filterType]);
    const newSelectionsState = cloneDeep(this.state.selections[filterType]);

    const targetInState = newOptionsState.find(opt => opt.name === name);
    targetInState.isSelected = isSelected;

    if (isSelected) {
      newSelectionsState.push(target);
    } else {
      const indexOnSelection = newSelectionsState.findIndex(opt => opt.name === name);
      newSelectionsState.splice(indexOnSelection, 1);
    }

    this.setState({
      options: { ...this.state.options, [filterType]: newOptionsState },
      selections: { ...this.state.selections, [filterType]: newSelectionsState },
    }, () => {
      if (!this.hasFiltersSelected()) {
        this.state.crosstalkHandle.clear();
        return;
      }
      const matchingRows = this.getFiltersMatchingRows();
      this.state.crosstalkHandle.set(matchingRows);
    });
  }

  filtersCount() {
    const count = Object.values(this.state.selections).reduce((res, filt) => res + filt.length, 0);
    return count > 0 ? count : "";
  }
  
  resetFilters() {
    this.setState({
      options: cloneDeep(this.parsedData),
      selections: shallowFiltersState(),
    }, () => {
      this.state.crosstalkHandle.clear();
    });
  }

  togglePane() {
    this.setState({ showPane: !this.state.showPane });
  }

  render() {
    return (
      <FilterContext.Provider value={this.state}>
        <div className="rscfilter">
          <button
            className="rscfilter__toggle"
            onClick={this.togglePane}
          >
            Filter
          <span
            className="rscfilter__count badge rounded-pill bg-primary"
          >
            {this.filtersCount()}
          </span>
          </button>
          <FilterFieldsPane
            isActive={this.state.showPane}
          />
        </div>
      </FilterContext.Provider>
    );
  }
}

Filter.propTypes = {
  data: PropTypes.object,
  crosstalkKey: PropTypes.array,
  crosstalkGroup: PropTypes.string,
}

export default Filter;
