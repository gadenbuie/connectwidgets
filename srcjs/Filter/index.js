import { reactWidget } from 'reactR';
import Filter from './Filter';

reactWidget('rsc_filter', 'output', { Filter });
