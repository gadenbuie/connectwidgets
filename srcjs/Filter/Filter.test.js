import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { range } from 'lodash';
import Filter from './Filter';
import FilterFieldsPane from './FilterFieldsPane';
import { FilterType } from './filterType';

const testFrameData = {
  // Filter component only needs this columns data.
  owner_username: [
    'jzcarpenter',
    'reporterfrank',
    'automata',
    'reporterfrank',
    'jzcarpenter',
    'reporterfrank',
    'autoreporter009',
    'brucebowtie',
  ],
  app_mode: [
    'rmd-static',   // Document
    'python-api',   // API
    'python-dash',  // Application
    'shiny',        // Application
    'rmd-static',   // Document
    'static',       // Pin
    'api',          // API
    'python-dash',  // Application
  ],
  content_category: [
    '',
    '',
    '',
    '',
    '',
    'pin',
    '',
  ],
  tags: [
    { name: ["vegetables", "carrot", "vitamin D"] },
    { name: "orange" },
    { name: ["fruits", "orange"] },
    { name: ["nutrition", "vitamin B", "vitamin D"] },
    { name: "watermelon" },
    { name: ["fruits", "strawberry"] },
    { name: ["other", "minerals", "zinc"] },
    { name: ["distros", "alpine"] },
  ],
};

/**
 * Expected filter options provided by parsed data
 * owner: ['automata' 'autoreporter009' 'brucebowtie' 'jzcarpenter' 'reporterfrank']
 * type: ['Document', 'API', 'Application', 'Pin']
 * tag: ['alpine' 'carrot' 'distros' 'fruits' 'minerals' 'nutrition' 'orange' 'other' 'strawberry' 'vegetables' 'vitamin B' 'vitamin D' 'watermelon' 'zinc']
 */

const mkWrapper = () => shallow(
  <Filter
    data={testFrameData}
    crosstalkKey={range(1, testFrameData.app_mode.length + 1)}
    crosstalkGroup="abcd-1234"
  />
);

describe('<Filter />', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders properly, setup state and crosstalk handle', () => {
    const wrapper = mkWrapper();
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(crosstalk.FilterHandle).toHaveBeenCalled();
    expect(crosstalk.FilterHandle().setGroup).toHaveBeenCalledWith('abcd-1234');
  });

  it('toggle pane', () => {
    const wrapper = mkWrapper();
    const button = () => wrapper.find('button');
    const pane = () => wrapper.find(FilterFieldsPane);

    expect(pane().props().isActive).toBe(false);

    button().simulate('click');
    expect(pane().props().isActive).toBe(true);

    button().simulate('click');
    expect(pane().props().isActive).toBe(false);
  });

  it('searchFieldOptions(), pre-filter options with the given term', () => {
    const wrapper = mkWrapper();
    const optsState = type => wrapper.state().stateOptionsFor(type);
    const ownerOptsState = () => optsState(FilterType.owner);
    const typeOptsState = () => optsState(FilterType.type);
    const tagsOptsState = () => optsState(FilterType.tag);

    wrapper.instance().searchFieldOptions(FilterType.owner, 'ter');
    expect(ownerOptsState().length).toBe(3);
    expect(ownerOptsState()[0].name).toBe('autoreporter009');
    expect(ownerOptsState()[1].name).toBe('jzcarpenter');
    expect(ownerOptsState()[2].name).toBe('reporterfrank');
    
    wrapper.instance().searchFieldOptions(FilterType.owner, 'auto');
    expect(ownerOptsState().length).toBe(2);
    expect(ownerOptsState()[0].name).toBe('automata');
    expect(ownerOptsState()[1].name).toBe('autoreporter009');

    wrapper.instance().searchFieldOptions(FilterType.owner, 'bruce');
    expect(ownerOptsState().length).toBe(1);
    expect(ownerOptsState()[0].name).toBe('brucebowtie');
    
    wrapper.instance().searchFieldOptions(FilterType.owner, '');
    expect(ownerOptsState().length).toBe(5);

    wrapper.instance().searchFieldOptions(FilterType.type, 'a');
    expect(typeOptsState().length).toBe(2);
    expect(typeOptsState()[0].name).toBe('API');
    expect(typeOptsState()[1].name).toBe('Application');

    wrapper.instance().searchFieldOptions(FilterType.type, 'doc');
    expect(typeOptsState().length).toBe(1);
    expect(typeOptsState()[0].name).toBe('Document');

    wrapper.instance().searchFieldOptions(FilterType.type, 'report');
    expect(typeOptsState().length).toBe(0);

    wrapper.instance().searchFieldOptions(FilterType.type, '');
    expect(typeOptsState().length).toBe(4);

    wrapper.instance().searchFieldOptions(FilterType.tag, 'vit');
    expect(tagsOptsState().length).toBe(2);
    expect(tagsOptsState()[0].name).toBe('vitamin B');
    expect(tagsOptsState()[1].name).toBe('vitamin D');

    wrapper.instance().searchFieldOptions(FilterType.tag, 'mine');
    expect(tagsOptsState().length).toBe(1);
    expect(tagsOptsState()[0].name).toBe('minerals');

    wrapper.instance().searchFieldOptions(FilterType.tag, '');
    expect(tagsOptsState().length).toBe(14);
  });

  it('updateSelection(), updates selected filters state and emits crosstalk event', () => {
    let expectedRows = [];
    const wrapper = mkWrapper();
    
    expectedRows = [2, 4, 6];
    wrapper.instance().updateSelection({
      name: 'reporterfrank',
      isSelected: true,
      filterType: FilterType.owner,
      applicableRows: [1, 3, 5],
    });
    expect(crosstalk.FilterHandle().set).toHaveBeenCalledWith(expectedRows);

    expectedRows = [2];
    wrapper.instance().updateSelection({
      name: 'API',
      isSelected: true,
      filterType: FilterType.type,
      applicableRows: [1, 6],
    });
    expect(crosstalk.FilterHandle().set).toHaveBeenCalledWith(expectedRows);

    expectedRows = [2, 7];
    wrapper.instance().updateSelection({
      name: 'autoreporter009',
      isSelected: true,
      filterType: FilterType.owner,
      applicableRows: [6],
    });
    expect(crosstalk.FilterHandle().set).toHaveBeenCalledWith(expectedRows);

    expectedRows = [2, 4, 6, 7];
    wrapper.instance().updateSelection({
      name: 'API',
      isSelected: false,
      filterType: FilterType.type,
      applicableRows: [1, 6],
    });
    expect(crosstalk.FilterHandle().set).toHaveBeenCalledWith(expectedRows);

    expectedRows = [7];
    wrapper.instance().updateSelection({
      name: 'reporterfrank',
      isSelected: false,
      filterType: FilterType.owner,
      applicableRows: [1, 3, 5],
    });
    expect(crosstalk.FilterHandle().set).toHaveBeenCalledWith(expectedRows);

    // Remove the last selected filter clears handle
    wrapper.instance().updateSelection({
      name: 'autoreporter009',
      isSelected: false,
      filterType: FilterType.owner,
      applicableRows: [6],
    });
    expect(crosstalk.FilterHandle().clear).toHaveBeenCalled();
  });

  it('resetFilters(), resets state and emits crosstalk clear', () => {
    const wrapper = mkWrapper();

    wrapper.instance().updateSelection({
      name: 'reporterfrank',
      isSelected: true,
      filterType: FilterType.owner,
      applicableRows: [1, 3, 5],
    });

    wrapper.instance().updateSelection({
      name: 'jzcarpenter',
      isSelected: true,
      filterType: FilterType.owner,
      applicableRows: [0, 4],
    });

    wrapper.instance().updateSelection({
      name: 'Document',
      isSelected: true,
      filterType: FilterType.type,
      applicableRows: [0, 4],
    });

    expect(wrapper.instance().hasFiltersSelected()).toBe(true);

    wrapper.instance().resetFilters();
    expect(wrapper.instance().hasFiltersSelected()).toBe(false);
    expect(crosstalk.FilterHandle().clear).toHaveBeenCalled();
  });
});
