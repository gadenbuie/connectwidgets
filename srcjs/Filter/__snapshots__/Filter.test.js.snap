// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Filter /> renders properly, setup state and crosstalk handle 1`] = `
<ContextProvider
  value={
    Object {
      "crosstalkHandle": Object {
        "clear": [MockFunction],
        "on": [MockFunction],
        "set": [MockFunction],
        "setGroup": [MockFunction] {
          "calls": Array [
            Array [
              "abcd-1234",
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
      },
      "dataKeys": Array [
        1,
        2,
        3,
        4,
        5,
        6,
        7,
        8,
      ],
      "options": Object {
        "owner": Array [
          Object {
            "applicableRows": Array [
              2,
            ],
            "filterType": "owner",
            "index": 0,
            "isSelected": false,
            "label": "automata",
            "name": "automata",
          },
          Object {
            "applicableRows": Array [
              6,
            ],
            "filterType": "owner",
            "index": 1,
            "isSelected": false,
            "label": "autoreporter009",
            "name": "autoreporter009",
          },
          Object {
            "applicableRows": Array [
              7,
            ],
            "filterType": "owner",
            "index": 2,
            "isSelected": false,
            "label": "brucebowtie",
            "name": "brucebowtie",
          },
          Object {
            "applicableRows": Array [
              0,
              4,
            ],
            "filterType": "owner",
            "index": 3,
            "isSelected": false,
            "label": "jzcarpenter",
            "name": "jzcarpenter",
          },
          Object {
            "applicableRows": Array [
              1,
              3,
              5,
            ],
            "filterType": "owner",
            "index": 4,
            "isSelected": false,
            "label": "reporterfrank",
            "name": "reporterfrank",
          },
        ],
        "tag": Array [
          Object {
            "applicableRows": Array [
              7,
            ],
            "filterType": "tag",
            "index": 0,
            "isSelected": false,
            "label": "alpine",
            "name": "alpine",
          },
          Object {
            "applicableRows": Array [
              0,
            ],
            "filterType": "tag",
            "index": 1,
            "isSelected": false,
            "label": "carrot",
            "name": "carrot",
          },
          Object {
            "applicableRows": Array [
              7,
            ],
            "filterType": "tag",
            "index": 2,
            "isSelected": false,
            "label": "distros",
            "name": "distros",
          },
          Object {
            "applicableRows": Array [
              2,
              5,
            ],
            "filterType": "tag",
            "index": 3,
            "isSelected": false,
            "label": "fruits",
            "name": "fruits",
          },
          Object {
            "applicableRows": Array [
              6,
            ],
            "filterType": "tag",
            "index": 4,
            "isSelected": false,
            "label": "minerals",
            "name": "minerals",
          },
          Object {
            "applicableRows": Array [
              3,
            ],
            "filterType": "tag",
            "index": 5,
            "isSelected": false,
            "label": "nutrition",
            "name": "nutrition",
          },
          Object {
            "applicableRows": Array [
              1,
              2,
            ],
            "filterType": "tag",
            "index": 6,
            "isSelected": false,
            "label": "orange",
            "name": "orange",
          },
          Object {
            "applicableRows": Array [
              6,
            ],
            "filterType": "tag",
            "index": 7,
            "isSelected": false,
            "label": "other",
            "name": "other",
          },
          Object {
            "applicableRows": Array [
              5,
            ],
            "filterType": "tag",
            "index": 8,
            "isSelected": false,
            "label": "strawberry",
            "name": "strawberry",
          },
          Object {
            "applicableRows": Array [
              0,
            ],
            "filterType": "tag",
            "index": 9,
            "isSelected": false,
            "label": "vegetables",
            "name": "vegetables",
          },
          Object {
            "applicableRows": Array [
              3,
            ],
            "filterType": "tag",
            "index": 10,
            "isSelected": false,
            "label": "vitamin B",
            "name": "vitamin B",
          },
          Object {
            "applicableRows": Array [
              0,
              3,
            ],
            "filterType": "tag",
            "index": 11,
            "isSelected": false,
            "label": "vitamin D",
            "name": "vitamin D",
          },
          Object {
            "applicableRows": Array [
              4,
            ],
            "filterType": "tag",
            "index": 12,
            "isSelected": false,
            "label": "watermelon",
            "name": "watermelon",
          },
          Object {
            "applicableRows": Array [
              6,
            ],
            "filterType": "tag",
            "index": 13,
            "isSelected": false,
            "label": "zinc",
            "name": "zinc",
          },
        ],
        "type": Array [
          Object {
            "applicableRows": Array [
              1,
              6,
            ],
            "filterType": "type",
            "index": 0,
            "isSelected": false,
            "label": "API",
            "name": "API",
          },
          Object {
            "applicableRows": Array [
              2,
              3,
              7,
            ],
            "filterType": "type",
            "index": 1,
            "isSelected": false,
            "label": "Application",
            "name": "Application",
          },
          Object {
            "applicableRows": Array [
              0,
              4,
            ],
            "filterType": "type",
            "index": 2,
            "isSelected": false,
            "label": "Document",
            "name": "Document",
          },
          Object {
            "applicableRows": Array [
              5,
            ],
            "filterType": "type",
            "index": 3,
            "isSelected": false,
            "label": "Pin",
            "name": "Pin",
          },
        ],
      },
      "resetFilters": [Function],
      "searchFieldOptions": [Function],
      "selections": Object {
        "owner": Array [],
        "tag": Array [],
        "type": Array [],
      },
      "showPane": false,
      "stateOptionsFor": [Function],
      "stateSelectionsFor": [Function],
      "updateSelection": [Function],
    }
  }
>
  <div
    className="rscfilter"
  >
    <button
      className="rscfilter__toggle"
      onClick={[Function]}
    >
      Filter
      <span
        className="rscfilter__count badge rounded-pill bg-primary"
      />
    </button>
    <FilterFieldsPane
      isActive={false}
    />
  </div>
</ContextProvider>
`;
